File: src/lib/payees.ts
import type { Payee } from '@/types';

/**
 * Picks the suggestions to offer for what has been typed so far.
 *
 * Prefix matches come first — typing "cor" should offer "Corner Market" before
 * "Tesco Corner" — and ties keep the order the server gave, which is
 * most-used-first. With nothing typed, the most-used entries are the suggestion.
 */
export function rankPayees(all: Payee[], typed: string, limit = 8): Payee[] {
	const query = typed.trim().toLowerCase();
	if (!query) return all.slice(0, limit);

	const startsWith = (entry: Payee) => entry.payee.toLowerCase().startsWith(query);

	return all
		.filter((entry) => entry.payee.toLowerCase().includes(query))
		.map((entry, index) => ({ entry, index }))
		.sort((a, b) => Number(startsWith(b.entry)) - Number(startsWith(a.entry)) || a.index - b.index)
		.slice(0, limit)
		.map(({ entry }) => entry);
}

/** True when the only match is what has already been typed, so the list is noise. */
export function isExhausted(matches: Payee[], typed: string): boolean {
	const query = typed.trim().toLowerCase();
	return matches.length === 1 && matches[0].payee.toLowerCase() === query;
}

File: app/src/main/java/dev/gavenda/yuuka/domain/Payees.kt
package dev.gavenda.yuuka.domain

import dev.gavenda.yuuka.data.model.Payee

/**
 * Picks the suggestions to offer for what has been typed so far.
 *
 * Prefix matches come first — typing "cor" should offer "Corner Market" before
 * "Tesco Corner" — and ties keep the order the server gave, which is
 * most-used-first. With nothing typed, the most-used entries are the suggestion.
 * Mirrors `src/lib/payees.ts`.
 */
fun rankPayees(all: List<Payee>, typed: String, limit: Int = 8): List<Payee> {
    val query = typed.trim().lowercase()
    if (query.isEmpty()) return all.take(limit)

    fun startsWith(entry: Payee) = entry.payee.lowercase().startsWith(query)

    return all
        .filter { it.payee.lowercase().contains(query) }
        .withIndex()
        .sortedWith(compareByDescending<IndexedValue<Payee>> { startsWith(it.value) }.thenBy { it.index })
        .take(limit)
        .map { it.value }
}

/** True when the only match is what has already been typed, so the list is noise. */
fun isExhausted(matches: List<Payee>, typed: String): Boolean {
    val query = typed.trim().lowercase()
    return matches.size == 1 && matches[0].payee.lowercase() == query
}
